### PR TITLE
Check warning text in tests

### LINKED ISSUE #1291: Prerelease warning made warning tests ineffective
Due to the fact that every prerelease version of a compiler emits a warning for every input, the tests that check for the existing of some warning are now ineffective. We should change these tests to check the actual warning text.

### PR DESCRIPTION
**Pull request by pirapira**
Trying to fix #1291.

### CODE REVIEW

**Comment by chriseth** [comment]
I would propose to use macros in the actual tests (so that the boost failure references the correct line) which check: error type, error message.

**Comment by chriseth** [comment]
Please also check that there is exactly one warning / exactly one error. If we have tests that report multiple errors, we might be able to fix them - if not, we can add another option to the macro.

**Comment by pirapira** [comment]
@chriseth "exactly one error" is split out to #1412 because this involves fixing 33 tests.

**Comment by pirapira** [comment]
@chriseth @axic I think this one is ready now.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 1]
```diff
@@ -177,7 +199,7 @@ BOOST_AUTO_TEST_CASE(double_stateVariable_declaration)
 					   "  uint256 variable;\n"
 					   "  uint128 variable;\n"
 					   "}\n";
-	BOOST_CHECK(expectError(text) == Error::Type::DeclarationError);
+	CHECK_ERROR(text, DeclarationError, "");
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:203]: One benefit I could see making these functions is that we could have two overloaded versions:

`check_error(text, DeclarationError);` vs `check_error(text, DeclarationError, "uint128 is invalid");`

**Comment by axic** [comment]
Looks great! We should also add expected error texts in a later PR wherever it fits.

File: libsolidity/interface/Exceptions.h [hunk 2]
```diff
@@ -57,6 +57,7 @@ class Error: virtual public Exception
 
 	Type type() const { return m_type; }
 	std::string const& typeName() const { return m_typeName; }
+	bool searchForSubstring(const std::string& _substr) const;
```
Comment (chriseth) [inline libsolidity/interface/Exceptions.h]: Declaration is still here.

File: test/libsolidity/ErrorCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+#include <test/libsolidity/ErrorCheck.h>
+#include <libdevcore/Exceptions.h>
+
+#include <string>
+
+bool dev::solidity::searchErrorMessage(Error const& _err, std::string const& _substr)
+{
+	if (const std::string* errorMessage = boost::get_error_info<dev::errinfo_comment>(_err))
```
Comment (chriseth) [inline test/libsolidity/ErrorCheck.cpp]: should be `std::string const*` - you can also add a `using namespace std;` above and remove the `std`

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 4]
```diff
@@ -61,7 +62,7 @@ parseAnalyseAndReturnError(string const& _source, bool _reportWarnings = false,
 
 		SyntaxChecker syntaxChecker(errors);
 		if (!syntaxChecker.checkSyntax(*sourceUnit))
-			return make_pair(sourceUnit, std::make_shared<Error::Type const>(errors[0]->type()));
+			return make_pair(sourceUnit, errors[0]);
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: `errors.at(0)`?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 1 above)
Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:203]: Functions will report the wrong failure location, and I think we should check for a message everywhere anyway.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 5]
```diff
@@ -2094,51 +2117,51 @@ BOOST_AUTO_TEST_CASE(positive_integers_to_unsigned_out_of_bound)
...
+	CHECK_ERROR(sourceCode, TypeError, "");
 }
 
 BOOST_AUTO_TEST_CASE(integer_boolean_operators)
 {
 	char const* sourceCode1 = R"(
 		contract test { function() { uint x = 1; uint y = 2; x || y; } }
 	)";
-	BOOST_CHECK(expectError(sourceCode1) == Error::Type::TypeError);
+	BOOST_CHECK(expectError(sourceCode1).type() == Error::Type::TypeError);
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: Why don't you use the `CHECK_ERROR` macro here?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 5 above)
Comment (pirapira) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Actually there are more of these.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 4 above)
Comment (pirapira) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: Yes, that's better.

File: test/libsolidity/ErrorCheck.cpp (hunk 3 above)
Comment (pirapira) [reply test/libsolidity/ErrorCheck.cpp]: I followed these.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 4 above)
Comment (pirapira) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: I've changed this.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 5 above)
Comment (pirapira) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp]: I've replaced all straightforward cases.

**Review by chriseth — APPROVED** [review]

**Pull request merged**